### PR TITLE
roachpb: pretty printer for roachpb.Value

### PR DESCRIPTION
**Pull request by danhhz**
Used in V(2) logging in the sql layer.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/7519)

### CODE REVIEW

**Comment by tamird** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[roachpb/data.go, line 559 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRjOoStEIElV4pTyno:-KLRjOoStEIElV4pTynp:1941524044) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/roachpb/data.go#L559)):*
> ```Go
> // PrettyPrint returns the value in a human readable format.
> // e.g. `Put /Table/51/1/1/0 -> /TUPLE/2:Int/2/1:Int/3`
> func (v Value) PrettyPrint() string {
> ```

mind writing some tests for this?

---

*[roachpb/data.go, line 562 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRjDUwzhdCo4_ausTS:-KLRjDUwzhdCo4_ausTT:-1923054496) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/roachpb/data.go#L562)):*
> ```Go
> 	var buf bytes.Buffer
> 	t := v.GetTag()
> 	buf.WriteString(`/`)
> ```

https://godoc.org/bytes#Buffer.WriteRune

---

*[roachpb/data.go, line 569 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRjJFAaoqRm1t9LqmJ:-KLRjJFAaoqRm1t9LqmK:329598490) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/roachpb/data.go#L569)):*
> ```Go
> 	switch t {
> 	case ValueType_TUPLE:
> 		var s string
> ```

why does this need to be outside the loop? seems like you could put it right about the call that populates it

---

*[roachpb/data.go, line 573 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRjFFlBlO__8d0n78E:-KLRjFFlBlO__8d0n78F:-1923054496) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/roachpb/data.go#L573)):*
> ```Go
> 		for i := 0; len(b) > 0; i++ {
> 			if i != 0 {
> 				buf.WriteString("/")
> ```

https://godoc.org/bytes#Buffer.WriteRune

---

*[roachpb/data.go, line 588 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRg3jrAqRBbwDXT7IL:-KLRg3jrAqRBbwDXT7IM:92902016) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/roachpb/data.go#L588)):*
> ```Go
> 		var i int64
> 		i, err = v.GetInt()
> 		fmt.Fprintf(&buf, "%d", i)
> ```

nit: strconv.Format* throughout for teh speed

---

*[roachpb/data.go, line 594 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRgEAx5gJIgIcMv0Y3:-KLRgEAx5gJIgIcMv0Y4:1491911485) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/roachpb/data.go#L594)):*
> ```Go
> 		fmt.Fprintf(&buf, "%f", f)
> 	case ValueType_BYTES:
> 		// TODO(dan): Try to sniff out strings and protos.
> ```

strings are doable but protos are going to be tricky (read: impossible?)

---

*[roachpb/data.go, line 597 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRioXMCgt48Rfu-WRd:-KLRioXMCgt48Rfu-WRe:-407240901) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/roachpb/data.go#L597)):*
> ```Go
> 		var data []byte
> 		data, err = v.GetBytes()
> 		fmt.Fprintf(&buf, "%x", data)
> ```

https://godoc.org/encoding/hex#EncodeToString or https://godoc.org/encoding/hex#Encode ?

---

*[roachpb/data.go, line 614 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRj4UFkF38_LUBQUJN:-KLRj4UFkF38_LUBQUJO:-617848863) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/roachpb/data.go#L614)):*
> ```Go
> 	}
> 	if err != nil {
> 		return fmt.Sprintf("%s:<err:%s>", t.String(), err.Error())
> ```

no need for .Error()

also, in the error cases, won't this produce double output? let's be sure to test some error cases.

---

*[sql/rowwriter.go, line 625 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRhlRibjoCyMSLPuSO:-KLRhlRjzEvplk-Aj97w:1005652746) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/sql/rowwriter.go#L625)):*
> ```Go
> 			ru.key = keys.MakeFamilyKey(primaryIndexKey, uint32(family.ID))
> 			if log.V(2) {
> 				log.Infof("Put %s -> %v", ru.key, ru.marshalled[idx].PrettyPrint())
> ```

perhaps as follow-up work: we repeat this pattern a whole lot - we should probably introduce a batch wrapper to the sql layer that adds this logging to every operation.

---

*[util/encoding/encoding.go, line 1474 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRjllV8uWiw9j1qfh_:-KLRjllV8uWiw9j1qfha:595569097) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/util/encoding/encoding.go#L1474)):*
> ```Go
> 			return b, "", err
> 		}
> 		return b, fmt.Sprintf("%x", data), nil
> ```

https://godoc.org/encoding/hex#EncodeToString

---

*[util/encoding/encoding.go, line 1490 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRjnOtE0_L9kZ93uV-:-KLRjnOtE0_L9kZ93uV0:-1982275936) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/util/encoding/encoding.go#L1490)):*
> ```Go
> 		return b, d.String(), nil
> 	default:
> 		return b, "", errors.Errorf("unknown tag %q", typ)
> ```

technically not a tag. also, why %q here but %s in roachpb?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-:-KLRkRD-o1OBVUvYx_yu:1417088721)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[util/encoding/encoding.go, line 1438 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRm60243qEu7K5x1dI:-KLRm60243qEu7K5x1dJ:484350173) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/util/encoding/encoding.go#L1438)):*
> ```Go
> // decodable value in the provided byte slice, along with the remaining byte
> // slice after decoding.
> func PrettyPrintValueEncoded(b []byte) ([]byte, string, error) {
> ```

This needs some tests as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519)*

File: roachpb/data.go [hunk 1]
```diff
@@ -554,6 +554,68 @@ func (v Value) computeChecksum(key []byte) uint32 {
 	return sum
 }
 
+// PrettyPrint returns the value in a human readable format.
+// e.g. `Put /Table/51/1/1/0 -> /TUPLE/2:Int/2/1:Int/3`
```
Comment (tschottdorf) [inline roachpb/data.go]: Can you explain the value format here? I can only guess what `2:Int/2/1:Int/3` is.

**Comment by paperstreet** [comment]
Review status: 0 of 5 files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[roachpb/data.go, line 558 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLR_y0RMxAYhJXzbx3_-r1-558:-KLSEbOZ7sMuDFbcbGvU:-843783959) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/roachpb/data.go#L558)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Can you explain the value format here? I can only guess what `2:Int/2/1:Int/3` is.
</details>
Great question. I was trying to match the level of verbosity that's currently used in key pretty-printing, which assumes you're quite familiar with our encoding. Happy to accept bikesheds for how better to print this

The first 2 is the column id (diff encoded, so the first column is id 2, the second is id 2+1)
The Int is the type of data
The second 2 is the value that was encoded

---

*[roachpb/data.go, line 559 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRjOoStEIElV4pTyno:-KLSDDCrcbzkPh9xRvxi:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/roachpb/data.go#L559)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> mind writing some tests for this?
</details>
Done.

---

*[roachpb/data.go, line 562 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRjDUwzhdCo4_ausTS:-KLRoopuURI0xd0wDcpv:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/roachpb/data.go#L562)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> https://godoc.org/bytes#Buffer.WriteRune
</details>
Done.

---

*[roachpb/data.go, line 569 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRjJFAaoqRm1t9LqmJ:-KLRot0DXuy4gIx89vEi:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/roachpb/data.go#L569)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why does this need to be outside the loop? seems like you could put it right about the call that populates it
</details>
Done.

---

*[roachpb/data.go, line 573 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRjFFlBlO__8d0n78E:-KLRotQEk2cw3RxEo-XZ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/roachpb/data.go#L573)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> https://godoc.org/bytes#Buffer.WriteRune
</details>
Done.

---

*[roachpb/data.go, line 588 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRg3jrAqRBbwDXT7IL:-KLSCsaqBKfpF_macmPY:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/roachpb/data.go#L588)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: strconv.Format* throughout for teh speed
</details>
Done.

---

*[roachpb/data.go, line 594 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRgEAx5gJIgIcMv0Y3:-KLRpjZfp0jo5iDJaa_6:431344485) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/roachpb/data.go#L594)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> strings are doable but protos are going to be tricky (read: impossible?)
</details>
Done. I was thinking we could use proto reflection, but realized we would need to guard against the case of something 

---

*[roachpb/data.go, line 597 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRioXMCgt48Rfu-WRd:-KLRqxJSPcnR6V9FTcYf:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/roachpb/data.go#L597)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> https://godoc.org/encoding/hex#EncodeToString or https://godoc.org/encoding/hex#Encode ?
</details>
Done.

---

*[roachpb/data.go, line 614 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRj4UFkF38_LUBQUJN:-KLSCiDvogDvjJYQTYmN:819845155) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/roachpb/data.go#L614)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> no need for .Error()
> 
> also, in the error cases, won't this produce double output? let's be sure to test some error cases.
</details>
I was being too clever in avoiding double output and ignoring buf in the error case. Added a comment.

Done and Done.

---

*[sql/rowwriter.go, line 625 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRhlRibjoCyMSLPuSO:-KLRs5H8c72-YFT90pH5:2133586090) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/sql/rowwriter.go#L625)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> perhaps as follow-up work: we repeat this pattern a whole lot - we should probably introduce a batch wrapper to the sql layer that adds this logging to every operation.
</details>
TODOed

---

*[util/encoding/encoding.go, line 1438 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRm60243qEu7K5x1dI:-KLSEZD4TPA3I_7HS3U7:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/util/encoding/encoding.go#L1438)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> This needs some tests as well.
</details>
Done.

---

*[util/encoding/encoding.go, line 1474 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRjllV8uWiw9j1qfh_:-KLSDENBFPclVWOpIxNr:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/util/encoding/encoding.go#L1474)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> https://godoc.org/encoding/hex#EncodeToString
</details>
Done.

---

*[util/encoding/encoding.go, line 1490 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLRjnOtE0_L9kZ93uV-:-KLSDFi9_G9Mj497CdHE:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/util/encoding/encoding.go#L1490)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> technically not a tag. also, why %q here but %s in roachpb?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519)*

**Comment by tamird** [comment]
:lgtm: again

---

Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[roachpb/data.go, line 600 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLSPriMdWohiwUGasZW:-KLSPriMdWohiwUGasZX:-595145788) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c6caa57939bfafe5dfacb7315ccdc8f7030e98/roachpb/data.go#L600)):*
> ```Go
> 		str := string(data)
> 		printable := true
> 		for _, c := range str {
> ```

ditto regarding https://godoc.org/strings#IndexFunc or https://godoc.org/bytes#IndexFunc

---

*[roachpb/data_test.go, line 1006 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLSQNMxyCIm53Ltje4R:-KLSQNMxyCIm53Ltje4S:-1153427081) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c6caa57939bfafe5dfacb7315ccdc8f7030e98/roachpb/data_test.go#L1006)):*
> ```Go
> 
> 	var timeValue Value
> 	timeValue.SetTime(time.Unix(1467216170, 5))
> ```

ditto regarding time.Date

---

*[roachpb/data_test.go, line 1036 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLSQc_BbETNIPlweqNA:-KLSQc_BbETNIPlweqNB:-2064785389) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c6caa57939bfafe5dfacb7315ccdc8f7030e98/roachpb/data_test.go#L1036)):*
> ```Go
> 		{decimalValue, "/DECIMAL/6.28"},
> 		{durationValue, "/DURATION/1m2d3ns"},
> 		{MakeValueFromBytes([]byte{1, 2, 3}), "/BYTES/010203"},
> ```

ditto re hex digits here

---

*[roachpb/data_test.go, line 1039 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLSQW728Q510Zot3E3Q:-KLSQW728Q510Zot3E3R:-1191211107) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c6caa57939bfafe5dfacb7315ccdc8f7030e98/roachpb/data_test.go#L1039)):*
> ```Go
> 		{MakeValueFromString("foo"), "/BYTES/foo"},
> 		{tupleValue, "/TUPLE/1:Int/8/2:Bytes/foo"},
> 		{errValue, "/<err:float64 value should be exactly 8 bytes: 1>"},
> ```

nit: space after `err: `

---

*[util/encoding/encoding.go, line 1484 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLSPe4UUPo78gV5hhNS:-KLSPe4UUPo78gV5hhNT:-1151583704) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c6caa57939bfafe5dfacb7315ccdc8f7030e98/util/encoding/encoding.go#L1484)):*
> ```Go
> 		str := string(data)
> 		printable := true
> 		for _, c := range str {
> ```

https://godoc.org/strings#IndexFunc might be useful here. in fact, you could use https://godoc.org/bytes#IndexFunc to potentially avoid the allocation.

---

*[util/encoding/encoding_test.go, line 1745 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLSP4j1Ua4lZ3o0PV7J:-KLSP4j1Ua4lZ3o0PV7K:-1946360320) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c6caa57939bfafe5dfacb7315ccdc8f7030e98/util/encoding/encoding_test.go#L1745)):*
> ```Go
> 		{EncodeFloatValue(nil, NoColumnID, 6.28), "6.28"},
> 		{EncodeDecimalValue(nil, NoColumnID, inf.NewDec(628, 2)), "6.28"},
> 		{EncodeTimeValue(nil, NoColumnID, time.Unix(1467216170, 5)), "2016-06-29T16:02:50.000000005Z"},
> ```

consider using https://godoc.org/time#Date instead of time.Unix here; it'll be easier to read.

---

*[util/encoding/encoding_test.go, line 1747 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLSPEG3vB6NfbE06duW:-KLSPEG3vB6NfbE06duX:-922194190) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c6caa57939bfafe5dfacb7315ccdc8f7030e98/util/encoding/encoding_test.go#L1747)):*
> ```Go
> 		{EncodeTimeValue(nil, NoColumnID, time.Unix(1467216170, 5)), "2016-06-29T16:02:50.000000005Z"},
> 		{EncodeDurationValue(nil, NoColumnID, duration.Duration{Months: 1, Days: 2, Nanos: 3}), "1m2d3ns"},
> 		{EncodeBytesValue(nil, NoColumnID, []byte{1, 2, 15}), "01020f"},
> ```

`[]byte{1, 2, 15}` -> `[]byte{0x1, 0x2, 0xF}` will be easier to visually inspect

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-:-KLSQgVScXg8dXSKNIaL:-494062255)*

**Comment by tschottdorf** [comment]
Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[roachpb/data.go, line 558 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLR_y0RMxAYhJXzbx3_-r1-558:-KLSX6KEVt792XUL9HJ2:-2104961455) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/roachpb/data.go#L558)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> Great question. I was trying to match the level of verbosity that's currently used in key pretty-printing, which assumes you're quite familiar with our encoding. Happy to accept bikesheds for how better to print this
> 
> The first 2 is the column id (diff encoded, so the first column is id 2, the second is id 2+1)
> The Int is the type of data
> The second 2 is the value that was encoded
> 
</details>
Thanks for the explanation, but what I was really asking is for that in the form of a comment in the source code for the next person wondering the same thing.
You're using the diff encoding to avoid large numbers when column families are mostly empty? I'm not sure that's useful enough (non-diffs might be much easier to parse?), but I might be missing the point.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519)*

**Comment by paperstreet** [comment]
Thanks for the reviews!

---

Review status: 1 of 5 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[roachpb/data.go, line 558 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLR_y0RMxAYhJXzbx3_-r1-558:-KLSYxYSF9Utnw8J3PHe:1520654447) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaefdc7e21c49ff3933ca0f0c662f7479039503d/roachpb/data.go#L558)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Thanks for the explanation, but what I was really asking is for that in the form of a comment in the source code for the next person wondering the same thing.
> You're using the diff encoding to avoid large numbers when column families are mostly empty? I'm not sure that's useful enough (non-diffs might be much easier to parse?), but I might be missing the point.
</details>
Aha! Sure, will do.

The diff encoding is used to keep varints small and save bytes. Upon further thought, I've changed it to print out both the diff and computed isd.

---

*[roachpb/data.go, line 600 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLSPriMdWohiwUGasZW:-KLSYttUm8q22_Mk5D0I:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c6caa57939bfafe5dfacb7315ccdc8f7030e98/roachpb/data.go#L600)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto regarding https://godoc.org/strings#IndexFunc or https://godoc.org/bytes#IndexFunc
</details>
Done.

---

*[roachpb/data_test.go, line 1006 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLSQNMxyCIm53Ltje4R:-KLSUQsVTghzfSlyBfUt:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c6caa57939bfafe5dfacb7315ccdc8f7030e98/roachpb/data_test.go#L1006)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto regarding time.Date
</details>
Done.

---

*[roachpb/data_test.go, line 1036 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLSQc_BbETNIPlweqNA:-KLSUS22_7fUkInGT493:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c6caa57939bfafe5dfacb7315ccdc8f7030e98/roachpb/data_test.go#L1036)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto re hex digits here
</details>
Done.

---

*[roachpb/data_test.go, line 1039 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLSQW728Q510Zot3E3Q:-KLSUZLioLYqof-Y7j4k:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c6caa57939bfafe5dfacb7315ccdc8f7030e98/roachpb/data_test.go#L1039)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: space after `err: `
</details>
Done.

---

*[util/encoding/encoding.go, line 1484 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLSPe4UUPo78gV5hhNS:-KLSY12mt5EagvCirv6O:-439788103) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c6caa57939bfafe5dfacb7315ccdc8f7030e98/util/encoding/encoding.go#L1484)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> https://godoc.org/strings#IndexFunc might be useful here. in fact, you could use https://godoc.org/bytes#IndexFunc to potentially avoid the allocation.
</details>
Used TrimIndexFunc so I didn't have to write a function just to invert the return value of IsPrint

---

*[util/encoding/encoding_test.go, line 1745 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLSP4j1Ua4lZ3o0PV7J:-KLSYw6IoDCXsUqvfluv:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c6caa57939bfafe5dfacb7315ccdc8f7030e98/util/encoding/encoding_test.go#L1745)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> consider using https://godoc.org/time#Date instead of time.Unix here; it'll be easier to read.
</details>
Done.

---

*[util/encoding/encoding_test.go, line 1747 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLSPEG3vB6NfbE06duW:-KLSU5DuJrMpAaNppP9x:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c6caa57939bfafe5dfacb7315ccdc8f7030e98/util/encoding/encoding_test.go#L1747)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `[]byte{1, 2, 15}` -> `[]byte{0x1, 0x2, 0xF}` will be easier to visually inspect
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-:-KLSa8doHHpWRKb_GVj2:-1063767586)*

**Comment by tamird** [comment]
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[roachpb/data_test.go, line 1038 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLSeOo1xAwY-hSNudaW:-KLSeOo1xAwY-hSNudaX:-1752491457) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f06cb2c621063454520bc813ee56222a48e1cd2/roachpb/data_test.go#L1038)):*
> ```Go
> 		{MakeValueFromBytes([]byte{0x1, 0x2, 0xF, 0xFF}), "/BYTES/01020fff"},
> 		{MakeValueFromString("foo"), "/BYTES/foo"},
> 		{tupleValue, "/TUPLE/1:1:Int/8/2:3:Bytes/foo"},
> ```

wondering if this wants some spaces

---

*[util/encoding/encoding.go, line 1484 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLSPe4UUPo78gV5hhNS:-KLSe1TBdcMJMHCoDwRt:267381851) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c6caa57939bfafe5dfacb7315ccdc8f7030e98/util/encoding/encoding.go#L1484)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> Used TrimIndexFunc so I didn't have to write a function just to invert the return value of IsPrint
</details>
ah yes, this is nice and clever (assuming that TrimLeftFunc doesn't allocate, which it shouldn't).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519)*

**Comment by paperstreet** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[roachpb/data_test.go, line 1038 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLSeOo1xAwY-hSNudaW:-KLShlTJykG29q8hoMN1:-2140075976) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f06cb2c621063454520bc813ee56222a48e1cd2/roachpb/data_test.go#L1038)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> wondering if this wants some spaces
</details>
The key pretty printer doesn't use any spaces, so I was trying to be consistent. Agree that it feels like this could be better, but I played with it for a bit and nothing concrete jumped out to me

---

*[util/encoding/encoding.go, line 1484 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519#-KLSPe4UUPo78gV5hhNS:-KLSeolxki5MkYQrkr3X:-1294310709) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c6caa57939bfafe5dfacb7315ccdc8f7030e98/util/encoding/encoding.go#L1484)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ah yes, this is nice and clever (assuming that TrimLeftFunc doesn't allocate, which it shouldn't).
</details>
```
   476	func TrimLeftFunc(s []byte, f func(r rune) bool) []byte {
   477		i := indexFunc(s, f, false)
   478		if i == -1 {
   479			return nil
   480		}
   481		return s[i:]
   482	}
   483	
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519)*

**Comment by tschottdorf** [comment]
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7519)*

**Pull request merged**